File: checks/src/test/java/com/cmgapps/lint/NullifyAnnotationDetectorTest.java
/*
 * Copyright (c) 2018. Christian Grach <christian.grach@cmgapps.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.cmgapps.lint;

import com.android.tools.lint.checks.infrastructure.LintDetectorTest;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Issue;

import java.util.Collections;
import java.util.List;

public class NullifyAnnotationDetectorTest extends LintDetectorTest {

    public void testMissingConstructorParameterAnnotation() {
        lint().files(
            java(
                "package test.pkg;\n" +
                    "public class Test {\n" +
                    "   public Test(int myInt, String myString){}\n" +
                    "}"
            ))
            .run()
            .expect("src/test/pkg/Test.java:3: Warning: Missing @NonNull or @Nullable [MissingNullifyAnnotation]\n" +
                "   public Test(int myInt, String myString){}\n" +
                "                          ~~~~~~~~~~~~~~~\n" +
                "0 errors, 1 warnings")
            .expectFixDiffs("Fix for src/test/pkg/Test.java line 2: Add @NonNull:\n" +
                "@@ -3 +3\n" +
                "-    public Test(int myInt, String myString){}\n" +
                "+    public Test(int myInt, @NonNull String myString){}\n" +
                "Fix for src/test/pkg/Test.java line 2: Add @Nullable:\n" +
                "@@ -3 +3\n" +
                "-    public Test(int myInt, String myString){}\n" +
                "+    public Test(int myInt, @Nullable String myString){}");
    }

    public void testConstructorParameterAnnotation() {
        lint().files(
            java(
                "package test.pkg;\n" +
                    "public class Test {\n" +
                    "   public Test(int myInt, @android.support.annotation.Nullable String myString){}\n" +
                    "}"
            )).run().expect("No warnings.");
    }

    public void testMissingFieldAnnotation() {
        lint().files(
            java(
                "package test.pkg;\n" +
                    "public class Test {\n" +
                    "   private String myString;\n" +
                    "   private int myInt;\n" +
                    "   private static final String constString = \"Test\";\n" +
                    "}"
            ))
            .run()
            .expect("src/test/pkg/Test.java:3: Warning: Missing @NonNull or @Nullable [MissingNullifyAnnotation]\n" +
                "   private String myString;\n" +
                "   ~~~~~~~~~~~~~~~~~~~~~~~~\n" +
                "0 errors, 1 warnings")
            .expectFixDiffs("Fix for src/test/pkg/Test.java line 2: Add @NonNull:\n" +
                "@@ -3 +3\n" +
                "-    private String myString;\n" +
                "+    @NonNull private String myString;\n" +
                "Fix for src/test/pkg/Test.java line 2: Add @Nullable:\n" +
                "@@ -3 +3\n" +
                "-    private String myString;\n" +
                "+    @Nullable private String myString;");
    }

    public void testFieldAnnotation() {
        lint().files(
            java(
                "package test.pkg;\n" +
                    "public class Test {\n" +
                    "   @android.support.annotation.NonNull private String myString;\n" +
                    "   private int myInt;\n" +
                    "   private static final String constString = \"Test\";\n" +
                    "}"
            )).run().expect("No warnings.");
    }

    public void testMissingMethodAnnotation() {
        lint().files(
            java(
                "package test.pkg;\n" +
                    "public class Test {\n" +
                    "   public void foo(int myInt, String myString){};\n" +
                    "}"
            ))
            .run()
            .expect("src/test/pkg/Test.java:3: Warning: Missing @NonNull or @Nullable [MissingNullifyAnnotation]\n" +
                "   public void foo(int myInt, String myString){};\n" +
                "                              ~~~~~~~~~~~~~~~\n" +
                "0 errors, 1 warnings")
            .expectFixDiffs("Fix for src/test/pkg/Test.java line 2: Add @NonNull:\n" +
                "@@ -3 +3\n" +
                "-    public void foo(int myInt, String myString){};\n" +
                "+    public void foo(int myInt, @NonNull String myString){};\n" +
                "Fix for src/test/pkg/Test.java line 2: Add @Nullable:\n" +
                "@@ -3 +3\n" +
                "-    public void foo(int myInt, String myString){};\n" +
                "+    public void foo(int myInt, @Nullable String myString){};");
    }

    public void testMethodAnnotation() {
        lint().files(java(
            "package test.pkg;\n" +
                "public class Test {\n" +
                "   public void foo(int myInt, @android.support.annotation.NonNull String myString){};\n" +
                "}"
        )).run().expect("No warnings.");
    }

    public void testMethodArrayMissingAnnotation() {
        lint().files(
            java(
                "package test.pkg;\n" +
                    "public class Test {\n" +
                    "   public void foo(int[] myInt, String[] myString){};\n" +
                    "}"
            ))
            .run()
            .expect("src/test/pkg/Test.java:3: Warning: Missing @NonNull or @Nullable [MissingNullifyAnnotation]\n" +
                "   public void foo(int[] myInt, String[] myString){};\n" +
                "                                ~~~~~~~~~~~~~~~~~\n" +
                "0 errors, 1 warnings")
            .expectFixDiffs("Fix for src/test/pkg/Test.java line 2: Add @NonNull:\n" +
                "@@ -3 +3\n" +
                "-    public void foo(int[] myInt, String[] myString){};\n" +
                "+    public void foo(int[] myInt, @NonNull String[] myString){};\n" +
                "Fix for src/test/pkg/Test.java line 2: Add @Nullable:\n" +
                "@@ -3 +3\n" +
                "-    public void foo(int[] myInt, String[] myString){};\n" +
                "+    public void foo(int[] myInt, @Nullable String[] myString){};");
    }

    public void testMethodArrayAnnotation() {
        lint().files(java(
            "package test.pkg;\n" +
                "public class Test {\n" +
                "   public void foo(int[] myInt, @android.support.annotation.Nullable String[] myString){};\n" +
                "}"
        )).run().expect("No warnings.");
    }

    public void testMissingReturnAnnotation() {
        lint().files(
            java(
                "package test.pkg;\n" +
                    "public class Test {\n" +
                    "   public int foo(){};\n" +
                    "   public String foo(){};\n" +
                    "}"
            ))
            .run()
            .expect("src/test/pkg/Test.java:4: Warning: Return type is missing @NonNull or @Nullable [MissingNullifyAnnotation]\n" +
                "   public String foo(){};\n" +
                "   ~~~~~~~~~~~~~~~~~~~~~\n" +
                "0 errors, 1 warnings")
            .expectFixDiffs("Fix for src/test/pkg/Test.java line 3: Add @NonNull:\n" +
                "@@ -4 +4\n" +
                "-    public String foo(){};\n" +
                "+    @NonNull public String foo(){};\n" +
                "Fix for src/test/pkg/Test.java line 3: Add @Nullable:\n" +
                "@@ -4 +4\n" +
                "-    public String foo(){};\n" +
                "+    @Nullable public String foo(){};");
    }

    public void testReturnAnnotation() {
        lint().files(java(
            "package test.pkg;\n" +
                "public class Test {\n" +
                "   public int foo(){};\n" +
                "   @android.support.annotation.NonNull\n" +
                "   public String foo(){};\n" +
                "}"
        )).run().expect("No warnings.");
    }

    public void testUninitializedFinalField() {
        lint().files(java(
            "package test.pkg;\n" +
                "public class Test {\n" +
                "   public final Foo foo;\n" +
                "}"
        )).run().expect("src/test/pkg/Test.java:3: Warning: Missing @NonNull or @Nullable [MissingNullifyAnnotation]\n" +
            "   public final Foo foo;\n" +
            "   ~~~~~~~~~~~~~~~~~~~~~\n" +
            "0 errors, 1 warnings");
    }

    public void testInitializedFinalField() {
        lint().files(java(
            "package test.pkg;\n" +
                "public class Test {\n" +
                "   public final Foo foo = new Foo();\n" +
                "}"
        )).run().expect("No warnings.");
    }

    public void testAnnotationDefinition() {
        lint().files(java(
            "package test.pkg;\n" +
                "public @interface Test {\n" +
                "   String value();\n" +
                "}"
        )).run().expect("No warnings.");
    }

    @Override
    protected Detector getDetector() {
        return new NullifyAnnotationDetector();
    }

    @Override
    protected List<Issue> getIssues() {
        return Collections.singletonList(NullifyAnnotationDetector.ISSUE);
    }
}

File: checks/src/test/java/com/cmgapps/lint/NullifyAnnotationDetectorTest.kt
/*
 * Copyright (c) 2018. Christian Grach <christian.grach@cmgapps.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.cmgapps.lint

import com.android.tools.lint.checks.infrastructure.TestFiles.java
import com.android.tools.lint.checks.infrastructure.TestLintTask.lint
import org.junit.Test

class NullifyAnnotationDetectorTest {

    @Test
    fun missingConstructorParameterAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    public Test(int myInt, String myString){}
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("""
                |src/test/pkg/Test.java:3: Warning: Missing @NonNull or @Nullable [MissingNullifyMethodAnnotation]
                |    public Test(int myInt, String myString){}
                |                           ~~~~~~~~~~~~~~~
                |0 errors, 1 warnings""".trimMargin()
            )
            .expectFixDiffs("""
                |Fix for src/test/pkg/Test.java line 2: Add @NonNull:
                |@@ -3 +3
                |-     public Test(int myInt, String myString){}
                |+     public Test(int myInt, @NonNull String myString){}
                |Fix for src/test/pkg/Test.java line 2: Add @Nullable:
                |@@ -3 +3
                |-     public Test(int myInt, String myString){}
                |+     public Test(int myInt, @Nullable String myString){}""".trimMargin()
            )
    }

    @Test
    fun correctConstructorParameterAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    public Test(int myInt, @android.support.annotation.Nullable String myString){}
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("No warnings.")
    }

    @Test
    fun missingFieldAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    private String myString;
                    |    private int myInt;
                    |    private static final String constString = "Test";
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("""
                |src/test/pkg/Test.java:3: Warning: Missing @NonNull or @Nullable [MissingNullifyFieldAnnotation]
                |    private String myString;
                |    ~~~~~~~~~~~~~~~~~~~~~~~~
                |0 errors, 1 warnings""".trimMargin()
            )
            .expectFixDiffs("""
                |Fix for src/test/pkg/Test.java line 2: Add @NonNull:
                |@@ -3 +3
                |-     private String myString;
                |+     @NonNull private String myString;
                |Fix for src/test/pkg/Test.java line 2: Add @Nullable:
                |@@ -3 +3
                |-     private String myString;
                |+     @Nullable private String myString;""".trimMargin()
            )
    }

    @Test
    fun correctFieldAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    @android.support.annotation.NonNull private String myString;
                    |    private int myInt;
                    |    private static final String constString = "Test";
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("No warnings.")
    }

    @Test
    fun correctFieldAndroidXAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    @androidx.annotation.Nullable private String myString;
                    |    private int myInt;
                    |    private static final String constString = "Test";
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("No warnings.")
    }

    @Test
    fun missingMethodAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    public void foo(int myInt, String myString){};
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("""
                |src/test/pkg/Test.java:3: Warning: Missing @NonNull or @Nullable [MissingNullifyMethodAnnotation]
                |    public void foo(int myInt, String myString){};
                |                               ~~~~~~~~~~~~~~~
                |0 errors, 1 warnings""".trimMargin()
            )
            .expectFixDiffs("""
                |Fix for src/test/pkg/Test.java line 2: Add @NonNull:
                |@@ -3 +3
                |-     public void foo(int myInt, String myString){};
                |+     public void foo(int myInt, @NonNull String myString){};
                |Fix for src/test/pkg/Test.java line 2: Add @Nullable:
                |@@ -3 +3
                |-     public void foo(int myInt, String myString){};
                |+     public void foo(int myInt, @Nullable String myString){};
                """.trimMargin())
    }

    @Test
    fun correctMethodAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    public void foo(int myInt, @android.support.annotation.NonNull String myString){};
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("No warnings.")
    }

    @Test
    fun correctMethodAndroidXAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    public void foo(int myInt, @androidx.annotation.NonNull String myString){};
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("No warnings.")
    }

    @Test
    fun missingMethodArrayAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    public void foo(int[] myInt, String[] myString){};
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("""
                |src/test/pkg/Test.java:3: Warning: Missing @NonNull or @Nullable [MissingNullifyMethodAnnotation]
                |    public void foo(int[] myInt, String[] myString){};
                |                                 ~~~~~~~~~~~~~~~~~
                |0 errors, 1 warnings""".trimMargin()
            )
            .expectFixDiffs("""
                |Fix for src/test/pkg/Test.java line 2: Add @NonNull:
                |@@ -3 +3
                |-     public void foo(int[] myInt, String[] myString){};
                |+     public void foo(int[] myInt, @NonNull String[] myString){};
                |Fix for src/test/pkg/Test.java line 2: Add @Nullable:
                |@@ -3 +3
                |-     public void foo(int[] myInt, String[] myString){};
                |+     public void foo(int[] myInt, @Nullable String[] myString){};
                """.trimMargin()
            )
    }

    @Test
    fun correctMethodArrayAnnotation() {
        lint()
            .files(
                java("""
                        |package test.pkg;
                        |public class Test {
                        |    public void foo(int[] myInt, @android.support.annotation.Nullable String[] myString){};
                        |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("No warnings.")
    }

    @Test
    fun missingReturnAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    public int foo(){};
                    |    public String foo(){};
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("""
                |src/test/pkg/Test.java:4: Warning: Return type is missing @NonNull or @Nullable [MissingNullifyMethodAnnotation]
                |    public String foo(){};
                |    ~~~~~~~~~~~~~~~~~~~~~
                |0 errors, 1 warnings""".trimMargin()
            )
            .expectFixDiffs("""
                |Fix for src/test/pkg/Test.java line 3: Add @NonNull:
                |@@ -4 +4
                |-     public String foo(){};
                |+     @NonNull public String foo(){};
                |Fix for src/test/pkg/Test.java line 3: Add @Nullable:
                |@@ -4 +4
                |-     public String foo(){};
                |+     @Nullable public String foo(){};
                """.trimMargin()
            )
    }

    @Test
    fun correctReturnAnnotation() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    public int foo(){};
                    |    @android.support.annotation.NonNull
                    |    public String foo(){};
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("No warnings.")
    }

    @Test
    fun uninitializedFinalField() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    public final Foo foo;
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("""
                |src/test/pkg/Test.java:3: Warning: Missing @NonNull or @Nullable [MissingNullifyFieldAnnotation]
                |    public final Foo foo;
                |    ~~~~~~~~~~~~~~~~~~~~~
                |0 errors, 1 warnings""".trimMargin()
            )
            .expectFixDiffs("""
                |Fix for src/test/pkg/Test.java line 2: Add @NonNull:
                |@@ -3 +3
                |-     public final Foo foo;
                |+     @NonNull public final Foo foo;
                |Fix for src/test/pkg/Test.java line 2: Add @Nullable:
                |@@ -3 +3
                |-     public final Foo foo;
                |+     @Nullable public final Foo foo;
                """.trimMargin()
            )
    }

    @Test
    fun initializedFinalField() {
        lint()
            .files(
                java("""
                    |package test.pkg;
                    |public class Test {
                    |    public final Foo foo = new Foo();
                    |}""".trimMargin()
                )
            )
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("No warnings.")
    }

    @Test
    fun testAnnotationDefinition() {
        lint()
            .files(
                java("""
                |package test.pkg;
                |public @interface Test {
                |   String value();
                |}""".trimMargin()
                ))
            .issues(*NullifyAnnotationDetector.getIssues())
            .run()
            .expect("No warnings.")
    }
}